src/lib/support/SerializableIntegerSet.cpp

```diff
@@ -0,0 +1,132 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "SerializableIntegerSet.h"
+
+namespace chip {
+
+const char * SerializableU64SetBase::SerializeBase64(char * buf, uint16_t & buflen)
```

Comment:
Using base64 serialization and string representation for persistent storage is quite inefficient (in many ways) on constrained embedded devices as it will take more flash for each non-volatile entry. Is this being used because the persistent store abstraction is String key / String value?

Reply:
Yes, the main need is due to string key/value pairs.
The current requirement for it is for the controller applications.

Reply:
Could the impl of the store on phone-based controller apps not just always take the buffer and do the serialization/deserialization in their storage layer? Then the code of multiple modules can agree on the actual structural serialization, using lowest common denominator (fixed key, fixed bag of bytes) method that is easy to map on embedded, and the actual mapping between a store/load request and the storage layer is separate? This would allow some entries to contain TLV, some others to contain representation such as you sparse array here, raw, and other could have structs, etc. It's all loaded/stored as byte spans, and indexed by key. I think it's OK to update the storage abstraction to decouple const char * storage ("string")  which has the horrible constraint of not allowing nul (\x00) bytes and always forcing a given serialization to string, even if none would be needed by the backing.

The keys can still be fixed strings because it's easy in the same storage layer "glue" to map fixed string --> scalar if there is a "scalar" key constraint such as some EEPROM layers that have numerical keys.

I think trying to overall follow a data storage abstraction that mandates strings as the stored entity may cause a scale problem as more and more things need to be stored (ACLs, certs, etc) and it will end-up defeating a whole bunch of critical size optimization work in the representations and buffer management to have to go through an intermediate base64 when the direct bytes could have been stored/loaded and possibly even indexed in place in flash for reads.

Reply:
Maybe this discussion can happen outside this PR, just so that we don't lose context of it. We can always add different encoding scheme to this class.